VSDiagnostics/VSDiagnostics/VSDiagnostics/Diagnostics/Exceptions/ExceptionThrownFromProhibitedContext/ExceptionThrownFromProhibitedContextAnalyzer.cs

```diff
@@ -0,0 +1,159 @@
+                VSDiagnosticsResources.ExceptionThrownFromDisposeAnalyzerMessage, Category, Severity, true);
+
+        private static DiagnosticDescriptor FinalizerRule
+            => new DiagnosticDescriptor(DiagnosticId.ExceptionThrownFromFinalizer,
+                VSDiagnosticsResources.ExceptionThrownFromFinalizerAnalyzerTitle,
+                VSDiagnosticsResources.ExceptionThrownFromFinalizerAnalyzerMessage, Category, Severity, true);
+
+        private static DiagnosticDescriptor GetHashCodeRule
+            => new DiagnosticDescriptor(DiagnosticId.ExceptionThrownFromGetHashCode,
+                VSDiagnosticsResources.ExceptionThrownFromGetHashCodeAnalyzerTitle,
+                VSDiagnosticsResources.ExceptionThrownFromGetHashCodeAnalyzerMessage, Category, Severity, true);
+
+        private static DiagnosticDescriptor EqualsRule
+            => new DiagnosticDescriptor(DiagnosticId.ExceptionThrownFromEquals,
+                VSDiagnosticsResources.ExceptionThrownFromEqualsAnalyzerTitle,
+                VSDiagnosticsResources.ExceptionThrownFromEqualsAnalyzerMessage, Category, Severity, true);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+            => ImmutableArray.Create(
+                ImplicitOperatorRule,
+                PropertyGetterRule,
+                StaticConstructorRule,
+                FinallyBlockRule,
+                EqualityOperatorRule,
+                DisposeRule,
+                FinalizerRule,
+                GetHashCodeRule,
+                EqualsRule);
+
+        public override void Initialize(AnalysisContext context) => context.RegisterSyntaxNodeAction(AnalyzeNode, SyntaxKind.ThrowStatement);
+
+        private void AnalyzeNode(SyntaxNodeAnalysisContext context)
+        {
+            // Since the current node is a throw statement there is no symbol to be found. 
+            // Therefore we look at whatever member is holding the statement (constructor, method, property, etc) and see what encloses that
+            var containingType = context.SemanticModel.GetEnclosingSymbol(context.Node.SpanStart).ContainingType;
+            var warningLocation = context.Node.GetLocation();
+
+            foreach (var ancestor in context.Node.Ancestors())
+            {
+                if (ancestor.IsKind(SyntaxKind.MethodDeclaration))
+                {
+                    var method = (MethodDeclarationSyntax) ancestor;
+                    var methodName = method.Identifier.ValueText;
+
+                    if (methodName == "Dispose")
+                    {
+                        var arity = method.ParameterList.Parameters.Count;
+                        context.ReportDiagnostic(Diagnostic.Create(DisposeRule, warningLocation, arity == 0 ? "Dispose()" : "Dispose(bool)", containingType.Name));
+                        return;
+                    }
+
+                    if (methodName == "GetHashCode")
```

Comment:
You could use the same technique I use to get these methods--by looking up the members of the type and comparing them.  That might be somewhat expensive for this analyzer, though, seeing how much it covers.